src/psyclone/psyir/transformations/scalarization_trans.py

```diff
@@ -0,0 +1,212 @@
+                    array_indices))
+            for index in flattened_indices:
+                sig, _ = index.get_signature_and_indices()
+                if var_accesses[sig].is_written():
+                    scalarizable = False
+                    break
+            if scalarizable:
+                potential_arrays.append(signature)
+
+        return potential_arrays
+
+    def _check_first_access_is_write(self, node, var_accesses, potentials):
+        potential_arrays = []
+
+        for signature in potentials:
+            if var_accesses[signature].is_written_first():
+                potential_arrays.append(signature)
+
+        return potential_arrays
+
+    def _check_valid_following_access(self, node, var_accesses, potentials):
+        potential_arrays = []
+
+        for signature in potentials:
+            # Find the last access of each signature
+            last_access = var_accesses[signature].all_accesses[-1].node
+            # Find the next access to this symbol
+            next_access = last_access.next_access()
+            # If we don't use this again then its valid
+            if next_access is None:
+                potential_arrays.append(signature)
+                continue
+            # If we do and the next_access has an ancestor IfBlock
+            # that isn't an ancestor of the loop then its not valid since
+            # we aren't tracking down what the condition-dependent next
+            # use really is.
+            if_ancestor = next_access.ancestor(IfBlock)
+
+            # If abs_position of if_ancestor is > node.abs_position
+            # its not an ancestor of us.
+            if (if_ancestor is not None and
+                    if_ancestor.abs_position > node.abs_position):
+                # Not a valid next_access pattern.
+                continue
+
+            # If next access is the LHS of an assignment, we need to
+            # check that it doesn't also appear on the RHS. If so its
+            # not a valid access
+            # I'm not sure this code is reachable
+#            if (isinstance(next_access.parent, Assignment) and
+#                next_access.parent.lhs is next_access and
+#                (next_access.next_access() is not None and
+#                 next_access.next_access().ancestor(Assignment) is
+#                 next_access.parent)):
+#                continue
+
+            # If next access is the RHS of an assignment then we need to
+            # skip it
+            ancestor_assign = next_access.ancestor(Assignment)
+            if (ancestor_assign is not None and
+                    ancestor_assign.lhs is not next_access):
+                continue
+
+            # If it has an ancestor that is a CodeBlock or Call or Kern
+            # then we can't guarantee anything, so we remove it.
+            if (next_access.ancestor((CodeBlock, Call, Kern))
+                    is not None):
+                continue
```

Comment:
I find hard to see what each of these are checking (without going to the tests), could you add a short code snipet showing the dependency that each is looking for, inlcuding the commented one.

Reply:
The commented one is removed because its not reachable with how the next_access/variable_access info gives results for assignments.